packages/smart-wallet/src/smartWallet.js

```diff
@@ -228,301 +235,288 @@ const behaviorGuards = {
+        helper.publishCurrentState();
+      },
+
+      publishCurrentState() {
+        const {
+          brandDescriptors,
+          currentPublishKit,
+          offerToUsedInvitation,
+          purseBalances,
+        } = this.state;
+        currentPublishKit.publisher.publish({
+          brands: [...brandDescriptors.values()],
+          purses: [...purseBalances.values()].map(a => ({
+            brand: a.brand,
+            balance: a,
+          })),
+          offerToUsedInvitation: mapToRecord(offerToUsedInvitation),
+          lastOfferId: ERROR_LAST_OFFER_ID,
+        });
+      },
+
+      /** @type {(desc: Omit<BrandDescriptor, 'displayInfo'>, purse: RemotePurse) => Promise<void>} */
+      async addBrand(desc, purseRef) {
+        const {
+          address,
+          brandDescriptors,
+          brandPurses,
+          paymentQueues,
+          updatePublishKit,
+        } = this.state;
+        // assert haven't received this issuer before.
+        const descriptorsHas = brandDescriptors.has(desc.brand);
+        const pursesHas = brandPurses.has(desc.brand);
+        assert(
+          !(descriptorsHas && pursesHas),
+          'repeated brand from bank asset subscription',
+        );
+        assert(
+          !(descriptorsHas || pursesHas),
+          'corrupted state; one store has brand already',
+        );
+
+        const [purse, displayInfo] = await Promise.all([
+          purseRef,
+          E(desc.brand).getDisplayInfo(),
+        ]);
+
+        // save all five of these in a collection (indexed by brand?) so that when
+        // it's time to take an offer description you know where to get the
+        // relevant purse. when it's time to make an offer, you know how to make
+        // payments. REMEMBER when doing that, need to handle every exception to
+        // put the money back in the purse if anything fails.
+        const descriptor = { ...desc, displayInfo };
+        brandDescriptors.init(desc.brand, descriptor);
+        brandPurses.init(desc.brand, purse);
+
+        const { helper } = this.facets;
+        // publish purse's balance and changes
+        E.when(
+          E(purse).getCurrentAmount(),
+          balance => helper.updateBalance(purse, balance, 'init'),
+          err =>
+            console.error(address, 'initial purse balance publish failed', err),
+        );
+        observeNotifier(E(purse).getCurrentAmountNotifier(), {
+          updateState(balance) {
+            helper.updateBalance(purse, balance);
+          },
+          fail(reason) {
+            console.error(address, `failed updateState observer`, reason);
+          },
+        });
+
+        updatePublishKit.publisher.publish({ updated: 'brand', descriptor });
+
+        // deposit queued payments
+        const payments = paymentQueues.has(desc.brand)
+          ? paymentQueues.get(desc.brand)
+          : [];
+        // @ts-expect-error xxx with DataOnly / FarRef types
```

Comment:
"xxx"?

Reply:
Acknowledgement of tech debt

https://github.com/Agoric/agoric-sdk/wiki/Coding-Style#code-comments